Initial documentation for the envelope encryption filter (#932)
kroxylicious/kroxylicious: docs/available-filters/envelope-encryption.adoc

:kms-api-javadoc: https://javadoc.io/doc/io.kroxylicious/kroxylicious-kms/latest

= Envelope Encryption

== What is it?

> A filter that provides a transparent encryption-at-rest solution for Apache Kafka.

Author: Maybe link to the use-case docs so the reader can understand more about the business problem being solved.

use a KMS in a highly available configuration.

Currently, there is a single KMS implementation which integrates with https://www.hashicorp.com/[HashiCorp Vault&#174;].
More KMS implementations are planned in the near future.

There is also a pluggable {kms-api-javadoc}/io/kroxylicious/kms/service/KmsService.html[public API] that may be used to
provide an alternative custom implementation.

== How to use the filter

There are three steps to using the filter.

1. Setting up the KMS.
2. Configuring the filter within Kroxylicious.
3. Establishing the encryption key(s) within the KMS that will be used to encrypt the topics.

These steps are described in the next sections.

=== Setting up the KMS

In order to set up the KMS provider ready to use with the filter, follow these provider specific steps.

==== HashiCorp Vault

The filter integrates with the https://developer.hashicorp.com/vault/docs/secrets/transit[HashiCorp Vault *transit
engine*].   The transit engine is not enabled by default.  It must be
https://developer.hashicorp.com/vault/docs/secrets/transit#setup[enabled] before it can be used with the filter.

The transit engine's path must be `/transit`.

> The filter requires the HashiCorp Vault endpoint and a https://developer.hashicorp.com/vault/docs/concepts/tokens[token].

Author: Really need to look more closely at the requirements of the token.  At the moment, I've only been experimenting with root tokens - which is unlikely configuration as the root token will be closely guarded.

These are used in the next section.

=== Filter Configuration

The filter is configured as part of the filter chain in the following way:

[source, yaml]
----
filters:
- type: EnvelopeEncryption                                        # <1>
  config:
    kms: KmsServiceName                                           # <2>
    kmsConfig:                                                    # <3>
      ..:
    selector: SelectorName                                        # <4>
    selectorConfig:                                               # <5>
      ..:
----
<1> The name of the filter. This must be `EnvelopeEncryption`.
<2> The KMS provider name.
<3> Object providing configuration understood by KMS provider.
<4> The key selector name.
<5> Object providing configuration understood by key selector.

==== KMS configuration

For the KMS configuration:

===== HashiCorp Vault

For HashiCorp Vault, the KMS configuration looks like this:

[source, yaml]
----
kms: VaultKmsService                                          # <1>
kmsConfig:
  vaultUrl: http://vault.vault.svc.cluster.local:8200         # <2>

>   vaultToken: myroottoken                                     # <3>

Author: @robobario we'll need to add your TLS work here.

----
selector: TemplateKekSelector                                 # <1>
selectorConfig:
  template: "key_${topicName}"                                # <2>
----
<1> The name of the KEK selector. This must be `TemplateKekSelector`.
<2> Template used to map for the topic name to the name of the key within the KMS.

=== Establishing the keys in the KMS

Use the management interface of the KMS to create the encryption keys.  The names(or aliases) of the encryption keys
must match the naming conventions established within the configuration of the KEK selector. If this is not done,
the filter will be unable to find the encryption key, and it won't be able to encrypt the produce requests.

For example, if using the `TemplateKekSelector` with the template `foo_$\{topicName}`, create a key for every topic that
is to be encrypted with the key name matching the topic name, prefixed by the string `foo`.

CAUTION: If the filter is unable to find a key for a given topic, the filter will pass through the records of the
produce request without encrypting the records.

==== HashiCorp Vault

Use either the HashiCorp UI or CLI to create AES-256 symmetric keys following your key naming convention.

> 

Author: Maybe add a CLI example to help users get going??

TIP: It is recommended to apply a key rotation policy to key as it is created.

=== Verifying that encryption is occurring

To verify that records sent to topics are indeed being encrypted, use `kafka-console-consumer` to consume the
records directly from the target Kafka Cluster.  Verify that encrypted text is seen rather than whatever plain text
that was sent by producer.  The record values seen will look something like this:

```
tradesvault:v1:+EfJ977UG1XkjI9yh7vxpgN2E1DKaIkDuxE+eCprVTKr+sskFuChcTe/KpR/c8ZDyP76W3itExmEzLOl����x)�Ũ�z�:S�������tБ��v���
```

> 

Author: The alternative verification is to look at the metrics.  I thought this was path of least resistence.

Reviewer: This is fine. It's more convincing proof, for a start.

:kms-api-javadoc: https://javadoc.io/doc/io.kroxylicious/kroxylicious-kms/latest
:encryption-api-javadoc: https://javadoc.io/doc/io.kroxylicious/kroxylicious-encryption

= Envelope Encryption

== What is it?

A filter that provides a transparent https://kroxylicious.io/use-cases/[encryption-at-rest solution] for Apache Kafka.

It works as follows:

1. The filter intercepts the requests sent from the producing Kafka Client containing records.  The records
   contained with the request are encrypted.  This happens before the records reach the Kafka Broker.
2. The filter intercepts the responses sent from the Kafka Broker to the consuming Kafka Client.  The records
   contained within the response are decrypted.  This happens before the records reach Kafka Client.

The entire process is transparent from the point of view of both Kafka Client and Kafka Broker.  Neither are
aware that the records are being encrypted.  Neither client nor broker have any access to the encryption keys.

The filter encrypts the kafka records using a symmetric encryption key. It is possible to use different encryption

> keys for different topics. The encryption keys are held within by a Key Management Service (KMS).  It is the user's

Reviewer: "The encryption keys" this bit isn't correct. The KEKs are, yes, but the encrypted DEKs are stored in Kafka. I know you're aiming for a simple explanation here. But I don't think it's helpful to give people an incorrect explanation. We should take the time and space to explain it properly, using pictures if that helps.

Author: I've added summary describing how envelope encryption is applied to the problem.  I've tried to keep it high level and reference the design for the details.

responsibility to provide the KMS service and pre-provision the encryption keys within it.

> 

Reviewer: This section is missing some other important details:

* We only encrypt the record values
* We pass through `null` unencrypted.

== Prerequisites

In order to use the filter, a compatible KMS (Key Management Service) must be provided.  The KMS is used to store
the encryption keys used to encrypt the records sent to the topic.  The service is used at runtime. If the KMS is
unavailable, it will become impossible to produce or consume records through the filter until the KMS service is
restored.  It is recommended to use a KMS in a highly available configuration.

Currently, there is a single KMS integration: https://www.hashicorp.com/[HashiCorp Vault&#174;].  More KMS integrations

> are planned soon.

Reviewer: I don't think we should promise anything, but certainly not with a promise of "soon".

```suggestion
are planned.
```

It is also possible for a user to implement their own KMS integration.  The implementation must implement the
{kms-api-javadoc}/io/kroxylicious/kms/service/KmsService.html[KMS public API] and make the implementation available on
the classpath with the service loader mechanism.

== How to use the filter

There are three steps to using the filter.

1. Setting up the KMS.

> 2. Configuring the filter within Kroxylicious.

Reviewer 2: Do we need to talk about setting up the classpath? The binary distribution only includes: 

```
kroxylicious-api-0.4.1.jar
kroxylicious-app-0.4.1.jar
kroxylicious-runtime-0.4.1.jar
```

so there's a bit of work to do to get the KMS/filter dependencies installed.

Reviewer 2: should we make a batteries included distro with all kroxy filters something we offer? otherwise we could create another encryption filter tarball containing.

```
caffeine-3.1.8.jar
kroxylicious-encryption-0.5.0-SNAPSHOT.jar
kroxylicious-kms-0.5.0-SNAPSHOT.jar
```

then the process would be take your clean Kroxy distro, unpack the encryption tarball from mavenCentral/github and download the KMS implementations from mavenCentral/github. A bit of a mission to explain.

Reviewer 2: or instruct them to curl those deps down out of central into their `libs` dir

Author: At this stage of  Kroxylicious's life, we have only one filter that has received serious work.  I think most users will be experimenting with it, rather than building their own.  I say with bundle it in the kroxylicious tgz for the time being.  We are already including the filters in the images.  Let's keep things simple.

Once we get to the point where we several filters, we can think alternative packaging strategies.  

WDYAT?   @SamBarker @tombentley

Reviewer: >  I say with bundle it in the kroxylicious tgz for the time being.

+1

3. Establishing the encryption key(s) within the KMS that will be used to encrypt the topics.

These steps are described in the next sections.

=== Setting up the KMS

In order to set up the KMS provider ready to use with the filter, follow these KMS provider specific steps.

==== HashiCorp Vault

The filter integrates with the https://developer.hashicorp.com/vault/docs/secrets/transit[HashiCorp Vault *transit
engine*].   Vault does not enable the transit engine by default.  It must be
https://developer.hashicorp.com/vault/docs/secrets/transit#setup[enabled] before it can be used with the filter.

The transit engine's path must be `/transit` (the default).

The filter's configuration requires:

* the HashiCorp Vault endpoint
* a https://developer.hashicorp.com/vault/docs/concepts/tokens[token] that will be used by the filter to read the
  encryption keys.

> 

Reviewer: Is there a simple command (e.g. `curl`) which the user can use at this point to verify that their Vault instance is ready to go?

> 

Author: There isn't the equivalent of a `vault token can-i`.  I'll have a think about the most succint way to to this.

> 

Author: I've added a shell script.

=== Filter Configuration

The filter is configured as part of the filter chain in the following way:

[source, yaml]
----
filters:
- type: EnvelopeEncryption                                        # <1>
  config:

>     kms: KmsServiceName                                           # <2>

Reviewer: Can we make this more obviously a placeholder?

    kmsConfig:                                                    # <3>
      ..:

>     selector: SelectorName                                        # <4>

Reviewer: Again, how does the reader know this is a placeholder, and not the actual value?

    selectorConfig:                                               # <5>
      ..:
----
<1> The name of the filter. This must be `EnvelopeEncryption`.
<2> The KMS provider name.

> <3> Object providing configuration understood by KMS provider.

Reviewer: ```suggestion
<3> Object providing configuration understood by the KMS provider.
```

<4> The key selector name.

> <5> Object providing configuration understood by key selector.

Reviewer: ```suggestion
<5> Object providing configuration understood by the key selector.
```

==== KMS configuration

For the KMS configuration:

===== HashiCorp Vault

For HashiCorp Vault, the KMS configuration looks like this:

[source, yaml]
----

> kms: VaultKmsService                                          # <1>

Reviewer: This makes me wonder if we really need the `Service` in the name. Or `Kms` for that matter. Just like with Filters and FilterFactories the person doing the configuring doesn't really care about the class name. It makes no sense to them why there's `Service` in the name. To them the KMS is just `HashicorpVault`

kmsConfig:
  vaultUrl: https://myvault:8200                              # <2>
  tls:                                                        # <3>
  vaultToken: mytoken                                         # <4>
----
<1> Name of the KMS provider. This must be `VaultKmsService`.
<2> Vault URL

> <3> (Optional) TLS trust configuration.

Reviewer: ```suggestion
<3> (Optional) TLS trust configuration. In this example no TLS is configured. 
```

<4> Vault Token

For TLS trust configuration, the filter accepts the same trust parameters as link:../deploying.adoc#_upstream_tls[Upstream TLS]
except the `PEM` store type is currently https://github.com/kroxylicious/kroxylicious/issues/933[not supported].

> ==== Key Encryption Key (KEK) selector configuration

Reviewer 2: Similar to what Tom was saying, talking about Key Encryption Key without defining it earlier (probably alongside defining DEK and a little more about how envelope encryption works) will be confusing. It's a funny name to see in isolation.

Maybe it needs to start with a `Concepts` section or `What is Envelope Encryption` before describing how to implement it with the filter.

> The role of the Key Encryption Key (KEK) selector to map from the topic to key that exists in the KMS.

Reviewer: ```suggestion
The role of the Key Encryption Key (KEK) selector to determine the key in the KMS which should be used to encrypt a given record.
```

Reviewer 2: I don't think we should talk about the KEK being used to encrypt a record. (I guess it is used in a roundabout way but it makes it sounds like it's used for the data encrypt)

NOTE: If the filter is unable to find a key for a given topic using the selector, the filter will pass through the
records of the produce request without encrypting them.

> The selector implementations are pluggable, so alternative implementations may be provided by implementing

Reviewer: Do we really want to be advertising this as a plugin API at this point? I.e. are we convinced that we can stand behind this as an API and aren't going to want to change it any time soon?

Author: Agreed

Author: Sentence removed.

the {encryption-api-javadoc}/io/kroxylicious/filter/encryption/KekSelectorService.html[KekSelectorService]
and making the implementation available on the classpath with the service loader mechanism.

===== Template KEK Selector

> The `TemplateKekSelector` maps from topic name to key name.  The template understand the substitution token

Reviewer: ```suggestion
The `TemplateKekSelector` maps from topic name to key name.  The template understands the substitution token
```

`$\{topicName}` which is replaced by the name of the topic.  It can be used to build key names
that include the topic name being encrypted.

Use the `$\{topicName}` is optional. It is possible to pass a literal string.  This will result in all topics being
encrypted using the same key.

[source, yaml]
----
selector: TemplateKekSelector                                 # <1>
selectorConfig:
  template: "key_${topicName}"                                # <2>
----
<1> The name of the KEK selector. This must be `TemplateKekSelector`.
<2> Template used to build the key name from the topic name.

=== Establishing the keys in the KMS

> Use the management interface of the KMS to create the encryption keys.  The names(or aliases) of the encryption keys

Reviewer: ```suggestion
Use the management interface of the KMS to create the encryption keys. The names (or aliases) of the encryption keys
```

aware that the records are being encrypted.  Neither client nor broker have any access to the encryption keys.

The filter encrypts the records using symmetric encryption keys.  The encryption technique employed is
known as *Envelope Encryption*, which is technique suited for encrypting large volumes of data in a efficient manner.
Envelope Encryption is described in the next section.

The filter integrates with a Key Management Service (KMS).  It is the Key Management Service (KMS) that has
ultimate responsibility for the safe storage of key material.  The role of the KMS is discussed later.

=== Envelope Encryption

Envelope Encryption is a technique described by https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-57pt1r5.pdf[NIST
Special Publication 800-57 Part 1 Revision 5]. It is the practice of encrypting the data with a data encryption key (DEK),
then wrapping (encrypting) the DEK with a key encryption key (KEK).  In this section we discuss how Envelope Encryption is
applied by the filter to encrypt Kafka records.  For more detail, refer to the {design-doc}[design].

> On the produce path, the filter is encrypting records.  The filter uses a selector to determine KEK is use. It then

Reviewer 2: ```suggestion
On the produce path, the filter is encrypting records.  The filter uses a selector to determine which KEK to use. It then
```

requests that the KMS generates a DEK for the KEK.  It is the DEK that is used to encrypt the record.  The original record
in the produce request is replaced by a *cipher record* which comprises the encrypted record, the encrypted DEK and some other
data.

On the fetch path, the filter is decrypting records.  The filter receives the *cipher record* from the Kafka Broker. The

> filter reverses process used to  construct the *cipher record* and uses the KMS to decrypt DEK.  The decrypted DEK is

Reviewer 2: ```suggestion
filter reverses the process used to  construct the *cipher record* and uses the KMS to decrypt the DEK.  The decrypted DEK is
```

used to decrypt the encrypted record.  The cipher record in the fetch response is then replaced with the decrypted record.

> On the produce path, the filter employs a DEK reuse strategy. This means that records sent to same topic will be encrypted

Reviewer 2: Should we mention that the strategy is currently per-client/per-connection?

using the same DEK (until a time-out or number of encryption limit is reached, whichever occurs first).  This prevents
key exhaustion and prevents excessive interactions with the KMS.

On the consume path, the filter employs an LRU strategy to keep recently encountered DEKs decrypted in memory. This
prevents excessive interactions with the KMS.

=== Role of the KMS

> The Key Management Service (KMS) is the *secure repository* for Key Management Keys (KEKs). The key material of the KEK

Reviewer 2: ```suggestion
The Key Management Service (KMS) is the *secure repository* for Key Encryption Keys (KEKs). The key material of the KEK
```

*never* leaves the confines of the KMS.  The KMS exposes cryptographic functions that operate against the KEKs stored
within it.

> The KMS is also the *source* of Data Encryption Keys (DEKs).  When the KMS generates a DEK for a given KEK, it returns

Reviewer 2: Should we explain that DEKs are not stored in the KMS? We use the KMS to generate the DEK but from then on it lives with the data.

Author: In the section before I already have:

"The original record in the produce request is replaced by a *cipher record* which comprises the encrypted record, the encrypted DEK and some other data."

so we were already explaining it, but I've added an extra sentence to this section too.

must implement the{kms-api-javadoc}/io/kroxylicious/kms/service/KmsService.html[KMS public API] and make the
implementation available on the classpath with the service loader mechanism.

IMPORTANT: It is recommended to use a KMS in a highly available configuration.

=== Key rotation

Key rotation is the practice of periodically replacing cryptographic keys with new ones.  Using key rotation is
considered cryptographic best-practice.

The filter allows for the rotation of KEKs within the KMS. When a KEK is rotated, the new key material will be applied
to newly produced records. Existing records (which were encrypted with older versions of the KEK) remain decryptable
as long as the previous KEK version remains present in the KMS.

WARNING: If the previous KEK version is removed from the KMS, the records encrypted with that key version will become
un-consumable (that is, the fetch will fail). In this case, the consumer offset must be advanced beyond those records.

=== What exactly gets encrypted?

The filter currently encrypts only *record values*.  Record keys, headers, timestamps are not encrypted.

> Null record values produced are passed through unencrypted.  THis serves to allow tombstoning.

Reviewer 2: ```suggestion
Null record values produced are passed through unencrypted.  This serves to allow tombstoning.
```
